src/client/components/MainForm.tsx

```diff
@@ -42,23 +58,111 @@ export const MainForm = ({
+  recaptchaSiteKey,
+  setRecaptchaErrorMessage,
+  setRecaptchaErrorContext,
   hasGuardianTerms = false,
   onSubmitOverride,
-}: PropsWithChildren<Props>) => {
-  const hasTerms = useRecaptcha || hasGuardianTerms;
+}: PropsWithChildren<MainFormProps>) => {
+  const recaptchaEnabled = !!recaptchaSiteKey;
+  const hasTerms = recaptchaEnabled || hasGuardianTerms;
+
+  const registerFormRef = createRef<HTMLFormElement>();
+  const [recaptchaState, setRecaptchaState] =
+    useState<UseRecaptchaReturnValue>();
+
+  /**
+   * Executes the reCAPTCHA check.
+   * Prevents the form from submitting until the reCAPTCHA check is complete.
+   */
+  const handleSubmit = useCallback(
+    (event: React.FormEvent<HTMLFormElement>) => {
+      if (onSubmitOverride) {
+        onSubmitOverride(event);
+      }
+
+      if (recaptchaEnabled && !recaptchaState?.token) {
+        event.preventDefault();
+        recaptchaState?.executeCaptcha();
+      }
+    },
+    [onSubmitOverride, recaptchaEnabled, recaptchaState],
+  );
+
+  /**
+   * Submits the form once the reCAPTCHA check has been successfully completed.
+   */
+  useEffect(() => {
+    if (recaptchaEnabled) {
+      const registerFormElement = registerFormRef.current;
+      if (recaptchaState?.token) {
+        registerFormElement?.submit();
+      }
+    }
+  }, [
+    recaptchaEnabled,
+    recaptchaState,
+    recaptchaState?.token,
+    registerFormRef,
+  ]);
+
+  useEffect(() => {
+    if (recaptchaEnabled) {
+      // Determine is something went wrong with the check.
+      const recaptchaCheckFailed =
+        recaptchaState?.error || recaptchaState?.expired;
+
+      // Used to show a more detailed reCAPTCHA error if
+      // the user has requested a check more than once.
+      const showErrorContext =
+        recaptchaCheckFailed && recaptchaState?.requestCount > 1;
+
+      // Default to generic reCAPTCHA error message.
+      // Show the retry message if the user has requested a check more than once.
+      const recaptchaErrorMessage = showErrorContext
+        ? CaptchaErrors.RETRY
+        : CaptchaErrors.GENERIC;
+
+      const recaptchaErrorContext = showErrorContext ? (
+        <DetailedRecaptchaError />
+      ) : undefined;
+
+      // Pass the error states back to the parent component if setters are provided.
+      if (recaptchaCheckFailed && setRecaptchaErrorMessage) {
+        setRecaptchaErrorMessage(recaptchaErrorMessage);
+      }
+      if (showErrorContext && setRecaptchaErrorContext) {
+        setRecaptchaErrorContext(recaptchaErrorContext);
+      }
+    }
+  }, [
+    recaptchaEnabled,
+    recaptchaState,
+    recaptchaState?.error,
+    recaptchaState?.expired,
+    recaptchaState?.requestCount,
+    setRecaptchaErrorContext,
+    setRecaptchaErrorMessage,
+  ]);
 
   return (
     <form
       css={formStyles}
       method="post"
       action={formAction}
-      onSubmit={onSubmitOverride}
+      onSubmit={handleSubmit}
+      ref={registerFormRef}
```

Comment:
the name of this is probably not right here, as `registerFormRef` implies it's to do with the registration form, rather than a generic form

I suggest maybe just using `formRef` as the name

Reply:
nice catch, missed this!